### PR TITLE
Fix performance degradation: make message fields lazy

### PR DESCRIPTION
**Pull request by odersky**
Make fields `msg` and `explanation` in Message lazy and make
message containers strict.

That way, we can look inside a message container to find out what kind
of message we have without fear of forcing too much.

[EDIT]

This has turned into a large-scale refactoring of the error reporting chain. The aim was to make things more consistent, simpler, and more robust. We also fix some problems where <nonsensical> tags would make it into the reported message, which was a classic case of people working on paralellel solutions that did not compose correctly with each other.

One reason for the drop in linecount is that I chucked ErrorMessagesTest out. I had little else but nuisance value.

### CODE REVIEW

**Comment by odersky** [comment]
We should also update the advice how to write an error message to include the laziness. And be vigilant for a which that nothing strict slips back in. The rules are:

 - Every message container should have `msg` and `explanation` (if defined) as lazy vals. 
 - The `kind` field is a strict val.
 - There should be no other strict vals in a `Message`.

**Comment by nicolasstucki** [comment]
test performance with #fast please

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/Message.scala [hunk 1]
```diff
@@ -12,7 +12,7 @@ object Message {
     * not yet been ported to the new scheme. Comment out this `implicit def` to
     * see where old errors still exist
     */
-  implicit def toNoExplanation(str: String): Message =
+  implicit def toNoExplanation(str: => String): Message =
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/Message.scala]: It's worth noting that we cannot have implicit conversion from a by-name param when using Dotty's `Conversion` mechanism.

Comment (odersky) [reply compiler/src/dotty/tools/dotc/reporting/diagnostic/Message.scala]: Yes, I  noted that as well. That's a problem.

**Comment by smarter** [comment]
> Please check http://lamppc37.epfl.ch:8000/pull-8603-03-24-21.43.out for more information

I connected to EPFL vpn to check it out, but got back:

> File not found. :(

/cc @liufengyun

**Comment by liufengyun** [comment]
It's running now --- I accidentally remove the `staging` remote while fixing an issue.

**Comment by nicolasstucki** [comment]
test performance please

**Comment by allanrenucci** [comment]
Why not make all the message fields (`msg`, `kind`, `explanation`) be `def`s? To the best of my knowledge, they are evaluated only once when they are reported.

**Comment by odersky** [comment]
@allanrenucci Good point. I don't know whether there are situations where we want to read these fields more than once.

**Comment by liufengyun** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
@anatoliykmetyuk can you give it a quick review? Since it touches many things it would be good to get this in sooner rather than later. It might still have some weaknesses but I am sure it's better than what we had before.

**Pull request merged**